# 数据挖掘-预测贷款用户是否逾期
## Task1：数据分析（2天）

-------
### 1 &emsp;数据类型的分析

**数据说明：** 

这份数据集是金融数据（非原始数据，已经处理过了），我们要做的是预测贷款用户是否会逾期。

表格中 "status" 是结果标签：0表示未逾期，1表示逾期。

 - **导入宏包**

In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

import warnings
warnings.filterwarnings('ignore')

import matplotlib as mpl
mpl.rcParams['font.sans-serif']=[u'SimHei']
mpl.rcParams['axes.unicode_minus']=False

 - 导入数据

In [2]:
data = pd.read_csv('./data/data.csv', encoding='gbk')
data.head()

Unnamed: 0   custid                          trade_no bank_card_no  \
0           5  2791858  20180507115231274000000023057383          卡号1   
1          10   534047  20180507121002192000000023073000          卡号1   
2          12  2849787  20180507125159718000000023114911          卡号1   
3          13  1809708  20180507121358683000000388283484          卡号1   
4          14  2499829  20180507115448545000000388205844          卡号1   

   low_volume_percent  middle_volume_percent  \
0                0.01                   0.99   
1                0.02                   0.94   
2                0.04                   0.96   
3                0.00                   0.96   
4                0.01                   0.99   

   take_amount_in_later_12_month_highest  trans_amount_increase_rate_lately  \
0                                      0                               0.90   
1                                   2000                               1.28   
2                                      0                               1.00   
3                                   2000                               0.13   
4                                      0                               0.46   

   trans_activity_month  trans_activity_day  ...  loans_max_limit  \
0                  0.55               0.313  ...           2900.0   
1                  1.00               0.458  ...           3500.0   
2                  1.00               0.114  ...           1600.0   
3                  0.57               0.777  ...           3200.0   
4                  1.00               0.175  ...           2300.0   

   loans_avg_limit  consfin_credit_limit  consfin_credibility  \
0           1688.0                1200.0                 75.0   
1           1758.0               15100.0                 80.0   
2           1250.0                4200.0                 87.0   
3           1541.0               16300.0                 80.0   
4           1630.0                8300.0                 79.0   

   consfin_org_count_current  consfin_product_count  consfin_max_limit  \
0                        1.0                    2.0             1200.0   
1                        5.0                    6.0            22800.0   
2                        1.0                    1.0             4200.0   
3                        5.0                    5.0            30000.0   
4                        2.0                    2.0             8400.0   

   consfin_avg_limit  latest_query_day  loans_latest_day  
0             1200.0              12.0              18.0  
1             9360.0               4.0               2.0  
2             4200.0               2.0               6.0  
3            12180.0               2.0               4.0  
4             8250.0              22.0             120.0  

[5 rows x 90 columns]

 - 查看数据属性

In [3]:
data.info()

print()
print("共有数据集：", data.shape[0])

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4754 entries, 0 to 4753
Data columns (total 90 columns):
Unnamed: 0                                    4754 non-null int64
custid                                        4754 non-null int64
trade_no                                      4754 non-null object
bank_card_no                                  4754 non-null object
low_volume_percent                            4752 non-null float64
middle_volume_percent                         4752 non-null float64
take_amount_in_later_12_month_highest         4754 non-null int64
trans_amount_increase_rate_lately             4751 non-null float64
trans_activity_month                          4752 non-null float64
trans_activity_day                            4752 non-null float64
transd_mcc                                    4752 non-null float64
trans_days_interval_filter                    4746 non-null float64
trans_days_interval                           4752 non-null float64
regional_mobility

 - 数据类型

In [4]:
for i,name in enumerate(data.columns):
    name_sum = data[name].value_counts().shape[0]
    print("{:2}、{:40}      The number of types of features is：{}".format(i + 1, name, name_sum))

 1、Unnamed: 0                                    The number of types of features is：4754
 2、custid                                        The number of types of features is：4754
 3、trade_no                                      The number of types of features is：4754
 4、bank_card_no                                  The number of types of features is：1
 5、low_volume_percent                            The number of types of features is：40
 6、middle_volume_percent                         The number of types of features is：90
 7、take_amount_in_later_12_month_highest         The number of types of features is：166
 8、trans_amount_increase_rate_lately             The number of types of features is：782
 9、trans_activity_month                          The number of types of features is：84
10、trans_activity_day                            The number of types of features is：512
11、transd_mcc                                    The number of types of features is：41
12、trans_days_interval_filter      

 - 数据统计

In [5]:
data.describe()

Unnamed: 0        custid  low_volume_percent  middle_volume_percent  \
count   4754.000000  4.754000e+03         4752.000000            4752.000000   
mean    6008.414178  1.690993e+06            0.021806               0.901294   
std     3452.071428  1.034235e+06            0.041527               0.144856   
min        5.000000  1.140000e+02            0.000000               0.000000   
25%     3106.000000  7.593358e+05            0.010000               0.880000   
50%     6006.500000  1.634942e+06            0.010000               0.960000   
75%     8999.000000  2.597905e+06            0.020000               0.990000   
max    11992.000000  4.004694e+06            1.000000               1.000000   

       take_amount_in_later_12_month_highest  \
count                            4754.000000   
mean                             1940.197728   
std                              3923.971494   
min                                 0.000000   
25%                                 0.000000   
50%                               500.000000   
75%                              2000.000000   
max                             68000.000000   

       trans_amount_increase_rate_lately  trans_activity_month  \
count                        4751.000000           4752.000000   
mean                           14.160674              0.804411   
std                           694.180473              0.196920   
min                             0.000000              0.120000   
25%                             0.615000              0.670000   
50%                             0.970000              0.860000   
75%                             1.600000              1.000000   
max                         47596.740000              1.000000   

       trans_activity_day   transd_mcc  trans_days_interval_filter  ...  \
count         4752.000000  4752.000000                 4746.000000  ...   
mean             0.365425    17.502946                   29.029920  ...   
std              0.170196     4.475616                   22.722432  ...   
min              0.033000     2.000000                    0.000000  ...   
25%              0.233000    15.000000                   16.000000  ...   
50%              0.350000    17.000000                   23.000000  ...   
75%              0.480000    20.000000                   32.000000  ...   
max              0.941000    42.000000                  285.000000  ...   

       loans_max_limit  loans_avg_limit  consfin_credit_limit  \
count      4457.000000      4457.000000           4457.000000   
mean       3390.038142      1820.357864           9187.009199   
std        1474.206546       583.418291           7371.257043   
min           0.000000         0.000000              0.000000   
25%        2300.000000      1535.000000           4800.000000   
50%        3100.000000      1810.000000           7700.000000   
75%        4300.000000      2100.000000          11700.000000   
max       10000.000000      6900.000000          87100.000000   

       consfin_credibility  consfin_org_count_current  consfin_product_count  \
count          4457.000000                4457.000000            4457.000000   
mean             76.042630                   4.732331               5.227507   
std              14.536819                   2.974596               3.409292   
min               0.000000                   0.000000               0.000000   
25%              77.000000                   2.000000               3.000000   
50%              79.000000                   4.000000               5.000000   
75%              80.000000                   7.000000               7.000000   
max              87.000000                  18.000000              20.000000   

       consfin_max_limit  consfin_avg_limit  latest_query_day  \
count        4457.000000        4457.000000       4450.000000   
mean        16153.690823        8007.696881         24.112809   
std         14301.037628        5679.418585         37.725724   
min             0

**结论一：** 可以看到数据共有4754行，其中浮点型数据70列、整数型13列、object 型7列，**需要对 object 类型的数据进行处理才能建立模型。**

**结论二：** 数据在整体上是相当完整的，**除了 student_feature 这个标签有较多缺失值以外，其他数据仅有很小的缺失。**

**结论三：** **大部分数据特征存在拖尾现象，** 即特征的 max(min) 明显偏离其 mean ，需要对这些数据进行进一步的处理才能建立准确的模型。

**注意：** 有的列像 'first_transaction_time'，它的值其实是日期的形式类似 20130817 这样，**pandas把它认为是int型的**，其实不是，这也需要注意。

-------
### 2&emsp;无关特征的删除

显然，上述特征中存在一些与贷款用户是否会逾期无关的特征。

经分析，无关特征及字段如下：
 - 'Unnamed: 0'
 - 客户id：'custid'
 - 流水号：'trade_no'
 - 卡号：'bank_card_no'
 - 资源：'source'
 - 客户姓名：'id_name'
 - 最新查询时间：'latest_query_time'
 - 最新贷款时间：'loans_latest_time'

In [6]:
data = data.drop(['Unnamed: 0', 'custid', 'trade_no', 'bank_card_no', 'source', 'id_name', 'latest_query_time','loans_latest_time'], axis = 1)
data.head()

low_volume_percent  middle_volume_percent  \
0                0.01                   0.99   
1                0.02                   0.94   
2                0.04                   0.96   
3                0.00                   0.96   
4                0.01                   0.99   

   take_amount_in_later_12_month_highest  trans_amount_increase_rate_lately  \
0                                      0                               0.90   
1                                   2000                               1.28   
2                                      0                               1.00   
3                                   2000                               0.13   
4                                      0                               0.46   

   trans_activity_month  trans_activity_day  transd_mcc  \
0                  0.55               0.313        17.0   
1                  1.00               0.458        19.0   
2                  1.00               0.114        13.0   
3                  0.57               0.777        22.0   
4                  1.00               0.175        13.0   

   trans_days_interval_filter  trans_days_interval  regional_mobility  ...  \
0                        27.0                 26.0                3.0  ...   
1                        30.0                 14.0                4.0  ...   
2                        68.0                 22.0                1.0  ...   
3                        14.0                  6.0                3.0  ...   
4                        66.0                 42.0                1.0  ...   

   loans_max_limit  loans_avg_limit  consfin_credit_limit  \
0           2900.0           1688.0                1200.0   
1           3500.0           1758.0               15100.0   
2           1600.0           1250.0                4200.0   
3           3200.0           1541.0               16300.0   
4           2300.0           1630.0                8300.0   

   consfin_credibility  consfin_org_count_current  consfin_product_count  \
0                 75.0                        1.0                    2.0   
1                 80.0                        5.0                    6.0   
2                 87.0                        1.0                    1.0   
3                 80.0                        5.0                    5.0   
4                 79.0                        2.0                    2.0   

   consfin_max_limit  consfin_avg_limit  latest_query_day  loans_latest_day  
0             1200.0             1200.0              12.0              18.0  
1            22800.0             9360.0               4.0               2.0  
2             4200.0             4200.0               2.0               6.0  
3            30000.0            12180.0               2.0               4.0  
4             8400.0             8250.0              22.0             120.0  

[5 rows x 82 columns]

--------
### 3&emsp;数据类型转换

由于特征并不都是数值型特征，因此我们还需要对数据类型进行转换。

经过无关特征删除后，我们还存有一个 object 类型特征（城市等级：'reg_preference_for_trad'），其余均为数值型特征。

In [7]:
print(data.reg_preference_for_trad.unique())

['一线城市' '三线城市' '境外' '二线城市' '其他城市' nan]


 - 数据类型转换

In [8]:
for i in range(0, data.shape[0]):
    if data.reg_preference_for_trad[i] == '一线城市':
        data.reg_preference_for_trad[i] = 1
    elif data.reg_preference_for_trad[i] == '二线城市':
        data.reg_preference_for_trad[i] = 2
    elif data.reg_preference_for_trad[i] == '三线城市':
        data.reg_preference_for_trad[i] = 3
    elif data.reg_preference_for_trad[i] == '境外':
        data.reg_preference_for_trad[i] = 4
    elif data.reg_preference_for_trad[i] == '其他城市':
        data.reg_preference_for_trad[i] = 5

# 特征提取时可以采用 one-hot 编码改进

In [9]:
data = data.convert_objects(convert_numeric=True)
print(data.dtypes.value_counts())

float64    71
int64      11
dtype: int64


--------
### 4&emsp;处理缺失值

 - 删除缺失值较多的行列
 
 若某个特征的缺失值超过 30%，那么它会损失特征的关键信息。
 
 同样若某个样本的缺失值过多，那么它也不再具有统计学意义。
 
 因此我们对这样的行或列直接进行删除。

In [10]:
data = data.dropna(axis=1, thresh = 1000)#删除缺失值超过 70% 的列
data = data.dropna(axis=0, thresh = 75)#删除缺失值超过7个的行

 - 填充缺失值
 
 对于剩余的缺失值，这里采用 **平均值填充法** 进行填充。

In [11]:
data = data.fillna(data.mean())
data

low_volume_percent  middle_volume_percent  \
0                   0.01                   0.99   
1                   0.02                   0.94   
2                   0.04                   0.96   
3                   0.00                   0.96   
4                   0.01                   0.99   
5                   0.02                   0.98   
6                   0.02                   0.98   
8                   0.03                   0.65   
9                   0.01                   0.99   
10                  0.01                   0.99   
11                  0.01                   0.99   
12                  0.00                   0.73   
13                  0.01                   0.79   
14                  0.01                   0.89   
15                  0.01                   0.99   
16                  0.02                   0.98   
17                  0.02                   0.98   
18                  0.01                   0.99   
19                  0.01                   0.96   
20                  0.02                   0.98   
21                  0.02                   0.98   
22                  0.00                   0.59   
23                  0.01                   0.99   
24                  0.02                   0.78   
25                  0.40                   0.60   
26                  0.02                   0.96   
27                  0.01                   0.91   
28                  0.01                   0.99   
30                  0.02                   0.93   
31                  0.02                   0.98   
...                  ...                    ...   
4723                0.01                   0.96   
4724                0.02                   0.98   
4725                0.01                   0.99   
4726                0.00                   1.00   
4727                0.02                   0.98   
4728                0.06                   0.94   
4729                0.17                   0.83   
4730                0.01                   0.99   
4731                0.00                   0.58   
4732                0.11                   0.89   
4733                0.03                   0.97   
4734                0.00                   1.00   
4735                0.01                   0.61   
4736                0.05                   0.95   
4737                0.01                   0.99   
4738                0.02                   0.98   
4739                0.05                   0.62   
4740                0.07                   0.93   
4741                0.00                   0.83   
4742                0.02                   0.98   
4743                0.00                   1.00   
4744                0.01                   0.96   
4745                0.04                   0.96   
4747                0.02                   0.40   
4748                0.09                   0.91   
4749                0.05                   0.68   
4750                0.01                   0.99   
4751                0.00                   1.00   
4752                0.05                   0.95   
4753                0.01                   0.99   

      take_amount_in_later_12_month_highest  \
0                                         0   
1                                      2000   
2                                         0   
3                                      2000   
4                                         0   
5                                      2000   
6                                         0   
8                                         0   
9                                       500   
10                                        0   
11                                      500   
12                                     5000   
13                                        0   
14                                     4000   
15                                        0   
16                                        0   
17                                      300 

-------
### 5&emsp;数据集切分

In [12]:
from sklearn.model_selection import train_test_split

train_data, test_data = train_test_split(data, test_size=0.3, random_state=2018)
train_data.to_csv('./data/train_data.csv', index=False, header=True)
test_data.to_csv('./data/test_data.csv', index=False, header=True)